### PR TITLE
Unsteady CHT

### PR DESCRIPTION
**Pull request by oleburghardt**
## Proposed Changes
These are rather smaller changes and fixes so that the unsteady multizone framework can be used for CHT.

A test case together with a tutorial (mainly on how - and how easy it is - to turn a steady simulation into an unsteady one) will be added.

## Related Work
[Tutorial](https://github.com/su2code/su2code.github.io/pull/26)
[Trailer on Youtube ;-)](https://www.youtube.com/watch?v=MqN8GalSyzk)

@TobiKattmann @ScSteffen Please follow this so that we can set up Unsteady Adjoint CHT computations afterwards.

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [x] My contribution is commented and consistent with SU2 style.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: SU2_CFD/src/drivers/CMultizoneDriver.cpp [hunk 1]
```diff
@@ -147,6 +147,10 @@ void CMultizoneDriver::StartSolver() {
 
   driver_config->Set_StartTime(StartTime);
 
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetDelta_UnstTimeND(config_container[ZONE_0]->GetDelta_UnstTimeND());
+  }
+
```
Comment (oleburghardt) [inline SU2_CFD/src/drivers/CMultizoneDriver.cpp]: This is ugly and if someone has got the time right now, I'd like to have your suggestions what the 'right' approach might be.

File: SU2_CFD/src/drivers/CMultizoneDriver.cpp [hunk 2]
```diff
@@ -139,6 +139,27 @@ CMultizoneDriver::~CMultizoneDriver(void) {
...
+  /*--- Find out the minimum of all references times and then set each zone to this (same) value.
+   * (To ensure that all zones run synchronously in time, be it a dimensional or non-dimensionalized one.) ---*/
+
+  su2double Time_Ref = config_container[ZONE_0]->GetTime_Ref();
+
+  for (iZone = 1; iZone < nZone; iZone++) {
+    if ((config_container[iZone]->GetTime_Ref() < Time_Ref)
+        && (config_container[iZone]->GetKind_Solver() != FEM_ELASTICITY))
+      Time_Ref = config_container[iZone]->GetTime_Ref();
+  }
```
Comment (oleburghardt) [inline SU2_CFD/src/drivers/CMultizoneDriver.cpp]: Update: This is a lot better as is does not rely on a hard-coded value like ZONE_0 but still..
Anyway, somehow we need to set the same reference time in all zones before starting an unsteady run.

File: SU2_CFD/src/drivers/CMultizoneDriver.cpp (hunk 1 above)
Comment (oleburghardt) [reply SU2_CFD/src/drivers/CMultizoneDriver.cpp]: It's improved now by taking the minimum of all reference times as the reference time for all zones (not requiring any hard-coded values).

File: SU2_CFD/include/iteration_structure.hpp [hunk 3]
```diff
@@ -747,19 +732,18 @@ class CHeatIteration : public CIteration {
...
-               CSolver *****solver,
-               CNumerics ******numerics,
-               CConfig **config,
-               CSurfaceMovement **surface_movement,
-               CVolumetricMovement ***grid_movement,
-               CFreeFormDefBox*** FFDBox,
-               unsigned short val_iZone,
-               unsigned short val_iInst);
+   void Solve(COutput *output,
+              CIntegration ****integration,
```
Comment (pcarruscag) [inline SU2_CFD/include/iteration_structure.hpp]: this is one space out of alignment

Comment (pcarruscag) [reply SU2_CFD/include/iteration_structure.hpp]: void is 3 spaces from the margin :nerd_face:

**Comment by oleburghardt** [comment]
> Have you checked that "suggest-parenthesis" warning? This would be a good PR to solve that.

Are those displayed by `-Wall`? Just checked that in that case there are no such "suggest paranthesis" warnings.

File: SU2_CFD/include/iteration_structure.hpp [hunk 4]
```diff
@@ -747,19 +732,18 @@ class CHeatIteration : public CIteration {
    * \param[in] surface_movement - Surface movement classes of the problem.
    * \param[in] grid_movement - Volume grid movement classes of the problem.
    * \param[in] FFDBox - FFD FFDBoxes of the problem.
-   * \param[in] val_iZone - zone of the problem.
```
[the remark this answers is not in the document]
Comment (oleburghardt) [reply SU2_CFD/include/iteration_structure.hpp]: They are missing everywhere. I think one should add those consistently.

File: SU2_CFD/include/solvers/CHeatSolver.hpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 /*!
- * \file CHeatSolverFVM.hpp
- * \brief Headers of the CHeatSolverFVM class
- * \author O. Burghardt
+ * \file CHeatSolver.hpp
+ * \brief Headers of the CHeatSolver class
+ * \author F. Palacios, T. Economon
```
[the remark this answers is not in the document]
Comment (oleburghardt) [reply SU2_CFD/include/solvers/CHeatSolver.hpp]: It's a very minor correction. Don't know why, maybe for consistency. There are no real guidelines for that anyway.

File: SU2_CFD/src/drivers/CSinglezoneDriver.cpp [hunk 6]
```diff
@@ -280,19 +280,18 @@ bool CSinglezoneDriver::Monitor(unsigned long TimeIter){
...
 
 
-  /*--- Check whether the outer time integration has reached the final time ---*/
-  TimeConvergence = GetTimeConvergence();
 
   if (TimeDomain == YES) {
 
     /*--- Check whether the outer time integration has reached the final time ---*/
 
+    TimeConvergence = GetTimeConvergence();
```
[the remark this answers is not in the document]
Comment (oleburghardt) [reply SU2_CFD/src/drivers/CSinglezoneDriver.cpp]: @ScSteffen Please check

File: SU2_CFD/src/iteration_structure.cpp [hunk 7]
```diff
@@ -1207,12 +1195,79 @@ void CHeatIteration::Iterate(COutput *output,
...
+#else
+  StartTime = MPI_Wtime();
+#endif
+
+  /*--- Preprocess the solver ---*/
+  Preprocess(output, integration, geometry, solver, numerics, config,
+            surface_movement, grid_movement, FFDBox, val_iZone, INST_0);
+
+  /*--- For steady-state flow simulations, we need to loop over ExtIter for the number of time steps ---*/
+  /*--- However, ExtIter is the number of FSI iterations, so nIntIter is used in this case ---*/
```
[the remark this answers is not in the document]
Comment (oleburghardt) [reply SU2_CFD/src/iteration_structure.cpp]: I think `CHeatIteration` is not the best place to start with correcting such comments. E.g. `CFluidIteration` is the most important iteration class and practically one can adapt only from there to have everything aligned. But you're still right, so feel free to replace it if you like :-)

**Comment by pcarruscag** [comment]
I don't know what option that warning is included with, or if the behaviour is different across compiler versions.
I compile with with -Wall -Wextra, nevertheless please check the line I pointed to in your previous PR, I would gladly silence the warning but I don't know how that logic is supposed to work.

**Comment by oleburghardt** [comment]
Ok. I'm still getting no warnings even with `-Wall -Wextra`.
But just to be sure I added some extra paranthesis around boolean operations in the code lines that you have referenced in my former PR.

**Comment by pcarruscag** [comment]
:+1:  What compiler (and version) do you use?

**Comment by oleburghardt** [comment]
gcc 9.2.1 :)

**Comment by pcarruscag** [comment]
[Here](https://gcc.godbolt.org/z/GykphR) you go, Wall is enough.

**Comment by oleburghardt** [comment]
Good to know, thanks! :+1:  The problem must have been solved by someone else before then (the latest commit with extra paranthesis is not a correction in that regard).

File: SU2_CFD/include/solvers/CHeatSolver.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/include/solvers/CHeatSolver.hpp]: "If you wrote everything you are the author" than seems to be the rule no?

**Comment by pcarruscag** [comment]
Hmm, someone fixed it in this branch then, because develop still has it, anyway its fixed.

**Comment by oleburghardt** [comment]
:-) Ok. Right know I'm trying to fix a bug for writing out surface sensitivities for the CHT cases in the repo. (It's probably not specific to CHT but a more general multizone problem.)

It would fit quite well here, so please hold on a second with merging!

**Pull request merged**